transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+            hint,
+            memo_fn,
+            rng,
+        )?;
+
+        let (amount, blinding) = tx_out
+            .masked_amount
+            .get_value(&shared_secret)
+            .expect("TransactionBuilder created an invalid Amount");
+        let output_secret = OutputSecret { amount, blinding };
+
+        self.impose_tombstone_block_limit(pubkey_expiry);
+
+        self.outputs_and_secrets
+            .push((tx_out.clone(), output_secret));
+
+        let confirmation = TxOutConfirmationNumber::from(&shared_secret);
+
+        Ok((tx_out, confirmation))
+    }
+
+    /// Sets the tombstone block, clamping to smallest pubkey expiry value.
+    ///
+    /// # Arguments
+    /// * `tombstone_block` - Tombstone block number.
+    pub fn set_tombstone_block(&mut self, tombstone_block: u64) -> u64 {
+        self.tombstone_block = min(tombstone_block, self.fog_tombstone_block_limit);
+        self.tombstone_block
+    }
+
+    /// Reduce the fog_tombstone_block_limit value by the amount specified,
+    /// and propagate this constraint to self.tombstone_block
+    fn impose_tombstone_block_limit(&mut self, pubkey_expiry: u64) {
+        // Reduce fog tombstone block limit value if necessary
+        self.fog_tombstone_block_limit = min(self.fog_tombstone_block_limit, pubkey_expiry);
+        // Reduce tombstone_block value if necessary
+        self.tombstone_block = min(self.fog_tombstone_block_limit, self.tombstone_block);
+    }
+
+    /// Consume the builder and return the transaction.
+    pub fn build<RNG: CryptoRng + RngCore>(
+        mut self,
+        rng: &mut RNG,
+    ) -> Result<SignedContingentInput, TxBuilderError> {
+        if !self.block_version.signed_input_rules_are_supported()
+            || !self.block_version.mixed_transactions_are_supported()
+        {
+            return Err(TxBuilderError::BlockVersionTooOld(
+                *self.block_version,
+                *BlockVersion::THREE,
+            ));
+        }
+
+        if self.block_version > BlockVersion::MAX {
+            return Err(TxBuilderError::BlockVersionTooNew(
+                *self.block_version,
+                *BlockVersion::MAX,
+            ));
+        }
+
+        self.outputs_and_secrets
+            .sort_by(|(a, _), (b, _)| a.public_key.cmp(&b.public_key));
+
+        let (outputs, output_secrets): (Vec<TxOut>, Vec<_>) =
+            self.outputs_and_secrets.drain(..).unzip();
+
+        let input_rules = InputRules {
+            required_outputs: outputs,
+            max_tombstone_block: if self.tombstone_block == u64::max_value() {
+                0
+            } else {
+                self.tombstone_block
+            },
+        };
+
+        // Now we can create the mlsag
+        let mut tx_in = TxIn::from(&self.input_credentials);
+        tx_in.input_rules = Some(input_rules.clone());
+        let mut ring = SignableInputRing::from(self.input_credentials);
+        ring.input_rules = Some(input_rules.clone());
```

Comment:
Does `ring` (and by association `ring.input_rules`) actually get used anywhere here? It looks like the code is only reading specific fields of `ring`.